### PR TITLE
[stdlib] Bounds checks and tests for Unsafe*BufferPointer

### PR DESCRIPTION
**Pull request by austinzheng**
#### What's in this pull request?

Bounds checks for `Unsafe*Buffer`'s subscript, and tests for the subscript getter. Setter tests will be in a forthcoming commit.

The stdlib test suite passes. Running all tests now...

@moiseev @gribozavr 

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

Changes:
- Added bounds checks to Unsafe*BufferPointer's subscript getter and setter
- Added tests for Unsafe*BufferPointer's subscript getter

### CODE REVIEW

**Comment by moiseev** [comment]
@swift-ci Please test

File: validation-test/stdlib/UnsafeBufferPointer.swift.gyb [hunk 1]
```diff
@@ -4,6 +4,86 @@
...
+  /// PRECONDITION: `memory` must be allocated with space for
+  /// `SubscriptGetTest.elementCount` elements. 
+  func createUnsafeBufferPointer(from memory: UnsafeMutablePointer<OpaqueValue<Int>>) 
+    -> UnsafeBufferPointer<OpaqueValue<Int>> 
+  {
+    for i in SubscriptGetTest.start..<SubscriptGetTest.end {
+      memory[i] = OpaqueValue(i)  
+    }
+    return UnsafeBufferPointer(start: memory, 
+                               count: SubscriptGetTest.elementCount)
```
Comment (moiseev) [inline validation-test/stdlib/UnsafeBufferPointer.swift.gyb]: Minor nitpick: I don't think we indent things this way. I'd rather have both arguments on the new line indented 2 spaces...

File: validation-test/stdlib/UnsafeBufferPointer.swift.gyb [hunk 2]
```diff
@@ -4,6 +4,86 @@
...
+  }
+
+  /// Create an populate an `UnsafeMutableBufferPointer` for use with unit
+  /// tests.
+  /// PRECONDITION: `memory` must be allocated with space for
+  /// `SubscriptGetTest.elementCount` elements.
+  func createUnsafeMutableBufferPointer(from memory: UnsafeMutablePointer<OpaqueValue<Int>>)
```
Comment (moiseev) [inline validation-test/stdlib/UnsafeBufferPointer.swift.gyb]: These 2 functions: `createUnsafeBufferPointer` and `createUnsafeMutableBufferPointer` are GYBbable.

File: validation-test/stdlib/UnsafeBufferPointer.swift.gyb [hunk 3]
```diff
@@ -105,6 +185,27 @@ ${SelfName}TestSuite.test("badNilCount")
...
+  }
+  let elementCount = SubscriptGetTest.elementCount
+
+  var memory = UnsafeMutablePointer<OpaqueValue<Int>>(allocatingCapacity: elementCount)
+  let buffer = test.create${SelfName}(from: memory)
+  defer {
+    memory.deallocateCapacity(elementCount)
+  }
+
+  let slice = buffer[test.range]
```
Comment (moiseev) [inline validation-test/stdlib/UnsafeBufferPointer.swift.gyb]: Maybe it's just me, but I would move the `expectCrashLater` immediately above this line. To emphasize which particular call is going to crash.

**Comment by moiseev** [comment]
Also, since we have overloads for the subscript that accept all other range types (specifically `ClosedRange`), I think it's worth testing those, probably in a separate PR.

**Comment by austinzheng** [comment]
Oops! I'll fix that.

**Comment by austinzheng** [comment]
It's ready for retesting whenever you are, @moiseev

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by austinzheng** [comment]
There was a GYB typo, sorry! I fixed it. You may need to trigger CI again.

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by austinzheng** [comment]
Thanks, and sorry that this PR is proving so difficult to get merged :/. Not my best day.

**Comment by moiseev** [comment]
Np.

**Comment by austinzheng** [comment]
I'm going to refactor part of this code. I'll let you know when it's ready for CI.

**Comment by austinzheng** [comment]
@moiseev I rewrote the tests so they actually compile and pass. As a bonus, all four range types are now checked (the change was pretty simple).

**Comment by moiseev** [comment]
@swift-ci Please test and merge

**Comment by moiseev** [comment]
@austinzheng looks great. Let's wait for the CI. Thanks!

**Pull request merged**